system/src/tcp.c

```diff
@@ -0,0 +1,356 @@
+/****************************************************************************
+ *
+ * File:
+ *      tcp.c
+ *
+ * Description:
+ *      Client and server interface abstraction using TCP/IP
+ *
+ * Author:
+ *      David Stockhouse
+ *
+ * Revision 0.1
+ *      Last edited 03/04/2020
+ *
+ * Revision 0.2
+ *      Last edited 03/19/2020
+ *      Changed some function interfaces
+ *
+ ***************************************************************************/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdarg.h>
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <arpa/inet.h>
+
+#include "debuglog.h"
+
+#include "tcp.h"
+
+
+/**** Function TCPClientInit ****
+ *
+ * Opens and initializes a TCP socket as a client. Creates a socket but does
+ * not attempt to connect to server.
+ *
+ * Arguments: 
+ *      None
+ *
+ * Return value:
+ *      On success, returns file descriptor corresponding to opened socket
+ *      On failure, prints error message and returns a negative number 
+ */
+int TCPClientInit() {
+
+    int sock_fd;
+
+    // Create bare socket
+    sock_fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sock_fd == -1) {
+        logDebug(L_INFO, "%s: Failed to create client socket\n", strerror(errno));
+        return sock_fd;
+    }
+
+    // Configure socket options
+    // Leaving this as a reminder to investigate further, but not yet using
+    // int socketOption;
+    // setsockopt(sock_fd, SOL_SOCKET, SO_OPTION, &socketOption, sizeof(int));
+
+    // Return file descriptor for socket
+    return sock_fd;
+
+} // TCPClientInit()
+
+
+/**** Function TCPClientTryConnect ****
+ *
+ * Attempts to establish a connection to the server at ipAddr and port. If the 
+ * connection cannot be made (i.e. the server is not ready), it returns 
+ * immediately instead of blocking.
+ *
+ * Arguments: 
+ *      sock_fd - File descriptor of open client socket
+ *      ipAddr  - String containing the IPv4 address "XXX.XXX.XXX.XXX"
+ *      port    - Integer port number to establish connection
+ *
+ * Return value:
+ *      Returns value returned by connect, which set errno appropriately
+ */
+int TCPClientTryConnect(int sock_fd, char *ipAddr, int port) {
+
+    int rc;
+    struct sockaddr_in socketAddress;
+
+    // Exit on error if invalid pointer
+    if (ipAddr == NULL) {
+        return -1;
+    }
+
+    // Configure socket address
+    socketAddress.sin_family = AF_INET;
+    socketAddress.sin_port = htons(port);
```

Comment:
Still not 100% sure after looking at the comments where htons is defined what this is doing.